osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,51 +20,165 @@ public class Speed : OsuStrainSkill
         private const double pi_over_4 = Math.PI / 4;
         private const double pi_over_2 = Math.PI / 2;
 
-        protected override double SkillMultiplier => 1400;
-        protected override double StrainDecayBase => 0.3;
+        private const double rhythm_multiplier = 2.0;
+        private const int history_time_max = 3000; // 3 seconds of calculatingRhythmBonus max.
+
+        private double skillMultiplier => 1375;
+        private double strainDecayBase => 0.3;
+
+        private double currentTapStrain = 1;
+        private double currentMovementStrain = 1;
+        private double currentRhythm = 1;
+
         protected override int ReducedSectionCount => 5;
         protected override double DifficultyMultiplier => 1.04;
 
         private const double min_speed_bonus = 75; // ~200BPM
         private const double max_speed_bonus = 45; // ~330BPM
         private const double speed_balancing_factor = 40;
 
+        protected override int HistoryLength => 32;
+
         public Speed(Mod[] mods)
             : base(mods)
         {
         }
 
-        protected override double StrainValueOf(DifficultyHitObject current)
+        private bool isRatioEqual(double ratio, double a, double b)
+        {
+            return a + 15 > ratio * b && a - 15 < ratio * b;
+        }
+
+        /// <summary>
+        /// Calculates a rhythm multiplier for the difficulty of the tap associated with historic data of the current <see cref="OsuDifficultyHitObject"/>.
+        /// </summary>
+        private double calculateRhythmBonus(DifficultyHitObject current)
         {
             if (current.BaseObject is Spinner)
                 return 0;
 
-            var osuCurrent = (OsuDifficultyHitObject)current;
+            int previousIslandSize = -1;
+            double rhythmComplexitySum = 0;
+            int islandSize = 0;
 
-            double distance = Math.Min(single_spacing_threshold, osuCurrent.TravelDistance + osuCurrent.JumpDistance);
-            double deltaTime = Math.Max(max_speed_bonus, current.DeltaTime);
+            bool firstDeltaSwitch = false;
 
-            double speedBonus = 1.0;
-            if (deltaTime < min_speed_bonus)
-                speedBonus = 1 + Math.Pow((min_speed_bonus - deltaTime) / speed_balancing_factor, 2);
+            for (int i = Previous.Count - 1; i > 0; i--)
```

Comment:
This indexing scheme could be somewhat improved if `Previous[i]` and `Previous[i - 1]` were stored to variables - it would make the fact that this is concerned with note pairs and not individual notes more self-explanatory.

Reply:
Done.